### ISSUE TITLE
Team Building from Box during Battle

### ISSUE THREAD

**Issue by Siddapa** [Siddapa/optimal-pkmn-battler#5]
Before a `Battle` is initialized, the Pokemon for both `Side`s need to be passed. Since this project aims to build the most optimal team from a box of Pokemon in someone's save file, the player's side will be initialized to empty with the enemy side having their Pokemon given. As the `Battle` progresses, switches to Pokemon still in the box need to be accounted as well as the one's on the team. Essentially, this paradigm follows the below setup:
```
If presented with choices after a "battle.update"
    If player's team has all 6 slots filled
        Only allow normal switches to Pokemon on team
    else
        Add all box Pokemon to possible choices as new switches (reusing pkmn.Choice object)
        If box switch is followed in a child node
            Update Battle struct with new Pokemon and slot order
            Score lower than normal switching (emphasizes box compression)
```

Currently, **@pkmn/engine** doesn't support this since it's obviously not a feature intended on cartridge. Nonetheless, no the gen1 mechanics impact the state of non-active Pokemon so behavior should be as expected.

**Comment by Siddapa** [comment]
Because the box state is dependent on the choices chosen by child nodes from parent nodes, the state of the box will be varied across the scoring nodes. When accounting for the multithreading, the box state should be stored in a way where access collisions are avoided. The following are potential solutions to this issue:

1. Store the box state within each `DecisionNode` (greatly increases the memory cost of building our tree)
    - **Pros**
        1. Easy access with no overhead of computation
        2. `DecisionNode`s are truly resource independent
    - **Cons**
        1. Greatly increases memory cost (storing the full `Battle` is already expensive)
2. Store the progression of the box states in each `DecisionNode` which would be the Pokemon already within the team
    - **Pros**
        1. No extra memory (Original Box - Current Team = Current Box State)
        2. Also requires no memory sharing by keeping the box globaly immutable
    - **Cons**
        1. Requires function for computing box state at any given node (could be done on-need basis)

2nd options clearly feels the most promising in terms of success in implementation and prioritizing other goals.

**Comment by Siddapa** [comment]
Fixed with dbb4ca39aac76e1aa3a362d96cb35af851635342

Box switches are now stored as an optional `next_turn` since leaving them as normal switches gets confusing when having to re-reference those transitions during traversal/scoring. Important to remember that the `next_turn` transition for a `box_switch` is a modification of the previous `battle` to the next one, but it CAN'T be applied like a normal engine switch since the Pokemon to switch to doesn't exist within the engine's team state. Either add the new Pokemon to the team and apply the switch like normal to recreate the next `battle` (HEAVILY discouraged) or just use the next `battle`'s state (encouraged).

Future considerations:
1. Multithreading scoring nodes will require a global lock on the `Box` which may need some personal engineering depending on how locks are done in Zig
2. Eliminating Pokemon that have already been added to the team when generating nodes for all Pokemon in the `Box` is quite inefficient currently. It has an _m x n_ where _m_ = `# of Box Pokemon` and _n_ = `# of Pokemon on team`. Should ideally just be a scan through the `Box` once
